Source/loadsave.cpp

```diff
@@ -1677,20 +1681,43 @@ bool IsHeaderValid(uint32_t magicNumber)
 	return false;
 }
 
+// Returns the size of the hotkeys file with the number of hotkeys passed and if a header with the number of hotkeys is present in the file
+size_t HotkeysSize(size_t nHotkeys=NUM_HOTKEYS, bool header=1) {
+	return (sizeof(uint8_t)*header) + (nHotkeys * sizeof(int32_t)) + (nHotkeys * sizeof(int8_t) + sizeof(int32_t) + sizeof(int8_t));
+}
+
 void LoadHotkeys()
 {
 	LoadHelper file("hotkeys");
-	if (!file.IsValid())
-		return;
 
 	auto &myPlayer = Players[MyPlayerId];
+	size_t nHotkeys = 4;
 
-	for (auto &spellId : myPlayer._pSplHotKey) {
-		spellId = static_cast<spell_id>(file.NextLE<int32_t>());
+	// Checking if the save file has the old format with only 4 hotkeys
+	if (!file.IsValid(HotkeysSize(4,0))) {
+		// If it doesn't match exactly the size expected for 4 hotkeys, assume we have a new style format and read the number of hotkeys from the beginning of the file 
+		nHotkeys = file.NextLE<uint8_t>();
 	}
-	for (auto &spellType : myPlayer._pSplTHotKey) {
-		spellType = static_cast<spell_type>(file.NextLE<int8_t>());
+
+	// Read all hotkeys in the file
+	for (size_t i = 0; i < nHotkeys; i++) {
+		// Do not load hotkeys past the size of the spell types array, discard the rest
+		if (i < NUM_HOTKEYS) {
+			myPlayer._pSplHotKey[i] = static_cast<spell_id>(file.NextLE<int32_t>());
+		} else {
+			file.Skip<int8_t>();
+		}
 	}
+	for (size_t i = 0; i < nHotkeys; i++) {
+		// Do not load hotkeys past the size of the spells array, discard the rest
+		if (i < NUM_HOTKEYS) {
+			myPlayer._pSplTHotKey[i] = static_cast<spell_type>(file.NextLE<int8_t>());
+		} else { 
+			file.Skip<int8_t>();
+		}
+	}

```

Comment:
If `nHotkeys < NUM_HOTKEYS` then `_pSplHotKey`/`_pSplTHotKey` will have uninitialised data at the end of the arrays (either garbage or whatever hotkeys were set in the last save game). Should fill `_pSplHotKey` with `SPL_INVALID` like what happens in `CreatePlayer` since that array seems to be the first checked in use.

Reply:
Ok, I did that for _pSplHotKey, but as far as I can see, _pSplTHotKey is never initialized, though when looking around quickly, there are checks in panels/spell_list.cpp that seem to deal with invalid values.